### PR TITLE
Add option to use MKL Pardiso, reduce number of tests

### PR DESCRIPTION
**Pull request by ranjanan**

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -54,9 +54,13 @@ function single_ground_all_pairs(data::GraphData{T,V}, flags, cfg, log = true) w
...
     else
-        csinfo("Solver used: CHOLMOD")
+        s = cfg["solver"]
+        mklpardiso = false
+        str = s in CHOLMOD ? "CHOLMOD" : "MKL Pardiso"
+        s in MKLPARDISO && (mklpardiso = true)
+        csinfo("Solver used: $str")
         bs = parse(Int, cfg["cholmod_batch_size"])
-        _cholmod_solver_path(data, flags, cfg, log, bs)
+        cholmod_solver_path(data, flags, cfg, log, bs, mklpardiso)
```
Comment (ViralBShah) [inline src/core.jl]: The function name is a bit strange. Can we use dispatch to pick the right one?

**Comment by ViralBShah** [comment]
Do we always check `norm(Ax-b)` after solving a system to make sure we got a reasonable answer, for all the solver types?

**Comment by ranjanan** [comment]
Fixes #188

File: src/core.jl [hunk 2]
```diff
@@ -42,36 +43,57 @@ struct Shortcut{T}
...
         csinfo("Solver used: AMG accelerated by CG")
-        amg_solver_path(data, flags, cfg, log)
-    else
+        return AMGSolver()
+    elseif s in CHOLMOD
+        csinfo("Solver used: CHOLMOD")
+        bs = parse(Int, cfg["cholmod_batch_size"])
+        return CholmodSolver(bs)
+    elseif s in MKLPARDISO
         csinfo("Solver used: CHOLMOD")
```
Comment (ViralBShah) [inline src/core.jl]: Shouldn't this be `csinfo("Solver used: MKLPardiso")`?

Comment (ranjanan) [reply src/core.jl]: Yes, correct.

File: src/core.jl (hunk 1 above)
Comment (ranjanan) [reply src/core.jl]: Latest commit reorgs the code

**Comment by ranjanan** [comment]
No idea why Windows is failing when I'm running completely serial code. Looking at the log, it seems like there's some reordering and race conditions even though I'm running with only one process.

**Comment by ranjanan** [comment]
Ok, the purpose of doing these experiments was the establish the following. If I did this: 

1. First I run a problem in parallel and get the answer
2. I remove the processes using `rmprocs`
3. I immediately run another problem after with 1 process. 

Then the `pmap` seems to be causing some kind of "race condition" (even though it is not running in parallel) and affecting one of the variables, thereby changing the final answer and causing my test to fail.

**Comment by ViralBShah** [comment]
Maybe merge the pardiso part of the work here and the other PR that is open, and chase this issue which seems serious separately?

**Comment by ViralBShah** [comment]
Or alternately have a separate PR to fix the broken test first, and then merge these two PRs.

**Comment by ViralBShah** [comment]
@ranjanan If you can further describe the pmap related issue or point to it in the code, perhaps @tanmaykm may be able to help. Maybe worthwhile to do a call with him and show him what's happening?

I feel like it is worth stabilizing this codebase fully before introducing threading. Thoughts?

**Comment by ranjanan** [comment]
I spoke to @shashi and we concluded that there's a race condition. 

Currently, everytime we run Circuitscape, we run `addprocs` at the beginning depending on the value of `max_parallel` in the config file. We also create am Array of SharedArray's whose length is `nprocs()` so each process can safely accumulate into their own matrix. We then do the computation and then run `rmprocs(1:workers())` at the end to get rid of the workers. 

Now, the race condition happens because `rmprocs` from the _previous_ test hasn't actually deleted workers but `nprocs()` returns 1. In the _next_ text, when we create a single SharedArray (because `nprocs()` returns 1), we have multiple processes writing into a single SharedArray.

Perhaps this would just be fixed if I did `wait(rmprocs(...))` instead so I know for sure the worker processes are removed.

**Comment by ViralBShah** [comment]
So what do we need to do to get this in?

**Comment by vlandau** [comment]
Probably doesn't make sense....but is there any sort of gc to handle actually deleting remote workers? I had a weird issue where I needed to do GC.gc() in runtests.jl for Omniscape to remove a lock in order to be able to remove a directory. Long shot, and not a well-informed idea by any means, but thought I'd mention it just in case.

**Comment by ViralBShah** [comment]
Can we rebase this one on master?

**Comment by ViralBShah** [comment]
@vlandau This is the pardiso branch - but needs to be rebased to master. Perhaps @ranjanan can rebase it and you can try it out and see the benefit?

**Comment by ViralBShah** [comment]
Nice - tests are passing!

**Pull request merged**